File: resto-app/lib/models/product.dart
import '../config/api_config.dart';

class Product {
  final int id;
  final String nom;
  final String? description;
  final double prix;
  final String? image;
  final int categorieId;
  final String? categorieNom;
  final bool disponible;
  final bool actif;

  Product({
    required this.id,
    required this.nom,
    this.description,
    required this.prix,
    this.image,
    required this.categorieId,
    this.categorieNom,
    this.disponible = true,
    this.actif = true,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as int,
      nom: json['nom'] as String,
      description: json['description'] as String?,
      prix: (json['prix'] as num).toDouble(),
      // L'API retourne image_url directement, sinon utiliser image
      image: json['image_url'] as String? ?? json['image'] as String?,
      categorieId: json['categorie_id'] as int,
      categorieNom: json['categorie']?['nom'] as String?,
      disponible: json['disponible'] == 1 || json['disponible'] == true,
      actif: json['actif'] == 1 || json['actif'] == true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nom': nom,
      'description': description,
      'prix': prix,
      'image': image,
      'categorie_id': categorieId,
      'disponible': disponible,
      'actif': actif,
    };
  }

  String get imageUrl {
    if (image == null) return '';
    // Si l'image est déjà une URL complète, la retourner telle quelle
    if (image!.startsWith('http://') || image!.startsWith('https://')) {
      return image!;
    }
    // Sinon, construire l'URL avec l'URL de base du serveur depuis ApiConfig
    return '${ApiConfig.serverBaseUrl}/storage/$image';
  }
}

File: app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Support\Facades\Storage;

class Product extends Model
{
    protected $table = 'produits';

    protected $fillable = [
        'categorie_id',
        'nom',
        'description',
        'prix',
        'image',
        'disponible',
        'actif',
    ];

    protected $casts = [
        'categorie_id' => 'integer',
        'prix' => 'decimal:2',
        'disponible' => 'boolean',
        'actif' => 'boolean',
    ];

    /**
     * Un produit appartient à une catégorie
     */
    public function categorie(): BelongsTo
    {
        return $this->belongsTo(Category::class, 'categorie_id');
    }

    /**
     * Un produit peut être dans plusieurs commandes
     */
    public function commandes(): BelongsToMany
    {
        return $this->belongsToMany(Commande::class, 'commande_produit')
            ->withPivot('quantite', 'prix_unitaire', 'notes')
            ->withTimestamps();
    }

    /**
     * Obtenir l'URL complète de l'image
     */
    public function getImageUrlAttribute(): ?string
    {
        if (!$this->image) {
            return null;
        }

        // Si le chemin commence par "public/", on l'enlève
        $path = str_starts_with($this->image, 'public/') 
            ? substr($this->image, 7) // Enlève "public/"
            : $this->image;

        // Si on est dans un contexte API ou JSON, retourner une URL complète
        // Sinon retourner un chemin relatif pour le web
        $request = request();
        if ($request && ($request->is('api/*') || $request->expectsJson())) {
            // Pour l'API, utiliser le domaine de la requête actuelle pour le développement local
            // ou config('app.url') pour la production
            $host = $request->getHttpHost();
            $scheme = $request->getScheme();
            
            // Si c'est un domaine local (localhost, 127.0.0.1, ou .test, .local), utiliser le domaine de la requête
            // Sinon utiliser config('app.url') pour la cohérence en production
            if (str_contains($host, 'localhost') || 
                str_contains($host, '127.0.0.1') || 
                str_contains($host, '.test') || 
                str_contains($host, '.local')) {
                return $scheme . '://' . $host . '/storage/' . $path;
            }
            
            // Pour la production, utiliser config('app.url')
            $baseUrl = rtrim(config('app.url'), '/');
            return $baseUrl . '/storage/' . $path;
        }

        // Retourner un chemin relatif qui fonctionne avec le domaine actuel
        // Le navigateur résoudra automatiquement l'URL complète
        return '/storage/' . $path;
    }

    /**
     * Scope pour les produits disponibles
     */
    public function scopeDisponibles($query)
    {
        return $query->where('disponible', true);
    }

    /**
     * Scope pour les produits actifs
     */
    public function scopeActifs($query)
    {
        return $query->where('actif', true);
    }

    /**
     * Scope pour filtrer par catégorie
     */
    public function scopeOfCategorie($query, int $categorieId)
    {
        return $query->where('categorie_id', $categorieId);
    }

    /**
     * Vérifier si le produit est disponible
     */
    public function isDisponible(): bool
    {
        return $this->actif && $this->disponible;
    }
}
